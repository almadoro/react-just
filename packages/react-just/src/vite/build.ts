import path from "node:path";
import type { Plugin } from "vite";
import { getRegisterModuleIdFromPath } from "./utils/client";

type BuildOptions = { entry: string };

export default function build(options: BuildOptions): Plugin {
  const clientEntries: string[] = [];

  return {
    name: "react-just:build",
    apply: "build",
    config() {
      return {
        appType: "custom",
        builder: {
          sharedConfigBuild: true,
          async buildApp(builder) {
            await builder.build(builder.environments.server);
            // It is required that the server build is executed first to catch all
            // the client entries.
            await builder.build(builder.environments.client);
          },
        },
        environments: {
          server: {
            consumer: "server",
            define: { "process.env.NODE_ENV": JSON.stringify("production") },
            build: {
              outDir: "dist/server",
              rollupOptions: {
                input: { index: path.resolve(options.entry) },
                output: { entryFileNames: "index.js" },
              },
            },
          },
          client: {
            consumer: "client",
            define: { "process.env.NODE_ENV": JSON.stringify("production") },
            build: {
              minify: false,
              outDir: "dist/client",
              rollupOptions: {
                input: { index: ENTRY_MODULE_ID },
                output: { entryFileNames: "index.js" },
              },
            },
          },
        },
      };
    },
    moduleParsed(moduleInfo) {
      if (this.environment.name !== "server") return;

      const isClientFile = moduleInfo.meta.reactUseClient?.transformed;
      if (isClientFile) clientEntries.push(moduleInfo.id);
    },
    resolveId(id) {
      if (id === ENTRY_MODULE_ID) return RESOLVED_ENTRY_MODULE_ID;
    },
    load(id) {
      if (id === RESOLVED_ENTRY_MODULE_ID)
        return getClientEntryCode(clientEntries);
    },
  };
}

export const ENTRY_MODULE_ID = "/virtual:client-entry";

const RESOLVED_ENTRY_MODULE_ID = "\0" + ENTRY_MODULE_ID;

function getClientEntryCode(entries: string[]) {
  let code = `import { hydrateFromWindowFlight, registerModule } from "react-just/client";`;

  for (let i = 0; i < entries.length; i++) {
    const entryPath = entries[i];
    const moduleId = getRegisterModuleIdFromPath(entryPath);
    code += `import * as entry_${i} from "${entryPath}";`;
    code += `registerModule("${moduleId}", entry_${i});`;
  }

  code += "hydrateFromWindowFlight();";
  return code;
}
