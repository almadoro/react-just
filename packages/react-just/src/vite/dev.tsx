import path from "node:path";
import type {
  Connect,
  PluginOption,
  RunnableDevEnvironment,
  ViteDevServer,
} from "vite";
import {
  renderToFlightPipeableStream,
  renderToHtmlPipeableStream,
} from "../../types/server.node";

type DevOptions = { app: string; flightMimeType: string };

export default function dev(options: DevOptions): PluginOption {
  return {
    name: "react-just:dev",
    apply: "serve",
    configureServer(server) {
      return () =>
        server.middlewares.use(middleware(server, options.flightMimeType));
    },
    resolveId(id) {
      if (id === CLIENT_ENTRY_MODULE_ID) return RESOLVED_CLIENT_ENTRY_MODULE_ID;
      if (id === SERVER_ENTRY_MODULE_ID) return RESOLVED_SERVER_ENTRY_MODULE_ID;
    },
    load(id) {
      if (id === RESOLVED_CLIENT_ENTRY_MODULE_ID) return CLIENT_ENTRY;
      if (id === RESOLVED_SERVER_ENTRY_MODULE_ID)
        return getServerEntry(
          path.resolve(this.environment.config.root, options.app),
        );
    },
  };
}

const CLIENT_ENTRY_MODULE_ID = "/virtual:client-entry";

const RESOLVED_CLIENT_ENTRY_MODULE_ID = "\0" + CLIENT_ENTRY_MODULE_ID;

const HMR_CODE =
  // Taken from: https://github.com/vitejs/vite-plugin-react/blob/main/packages/common/refresh-utils.ts
  `import { injectIntoGlobalHook } from "/@react-refresh";` +
  `injectIntoGlobalHook(window);` +
  `window.$RefreshReg$ = () => {};` +
  `window.$RefreshSig$ = () => (type) => type;`;

const CLIENT_ENTRY =
  HMR_CODE +
  `import { hydrateFromWindowFlight } from "react-just/client";` +
  `hydrateFromWindowFlight();`;

const SERVER_ENTRY_MODULE_ID = "/virtual:server-entry";

const RESOLVED_SERVER_ENTRY_MODULE_ID = "\0" + SERVER_ENTRY_MODULE_ID;

type RenderToHtmlPipeableStream = typeof renderToHtmlPipeableStream;
type RenderToFlightPipeableStream = typeof renderToFlightPipeableStream;

function middleware(
  server: ViteDevServer,
  flightMimeType: string,
): Connect.NextHandleFunction {
  return async (req, res) => {
    const ssr = server.environments.ssr as RunnableDevEnvironment;

    const serverModule = await ssr.runner.import(SERVER_ENTRY_MODULE_ID);

    const { renderToFlightPipeableStream, renderToHtmlPipeableStream, App } =
      serverModule as {
        renderToFlightPipeableStream: RenderToFlightPipeableStream;
        renderToHtmlPipeableStream: RenderToHtmlPipeableStream;
        App: React.ComponentType;
      };

    const AppRoot = () => (
      <>
        <script async type="module" src={CLIENT_ENTRY_MODULE_ID} />
        <App />
      </>
    );

    if (req.headers.accept?.includes(flightMimeType)) {
      res.statusCode = 200;
      res.setHeader("content-type", flightMimeType);
      const flightStream = renderToFlightPipeableStream(<AppRoot />);
      flightStream.pipe(res);
      return;
    }

    res.statusCode = 200;
    res.setHeader("content-type", "text/html");
    const htmlStream = renderToHtmlPipeableStream(<AppRoot />);
    htmlStream.pipe(res);
  };
}

function getServerEntry(appPath: string) {
  return (
    `import { renderToFlightPipeableStream, renderToHtmlPipeableStream } from "react-just/server.node";` +
    `import App from "${path.resolve(appPath)}";` +
    `export { renderToFlightPipeableStream, renderToHtmlPipeableStream, App } `
  );
}
