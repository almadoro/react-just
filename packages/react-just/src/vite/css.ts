import { DevEnvironment, Plugin } from "vite";
import { isFlightEnvironment } from "./environments";
import { isCssModule } from "./utils";

export default function css(): Plugin {
  const flightEnvironments: DevEnvironment[] = [];

  return {
    name: "react-just:css",
    configureServer(server) {
      for (const environment of Object.values(server.environments)) {
        if (isFlightEnvironment(environment.name))
          flightEnvironments.push(environment);
      }
    },
    resolveId(id) {
      if (id === CSS_MODULES) return RESOLVED_CSS_MODULES;
    },
    load(id) {
      if (id === RESOLVED_CSS_MODULES)
        return getCssModulesCode(flightEnvironments);
    },
  };
}

export const CSS_MODULES = "/virtual:react-just/css-modules";
export const RESOLVED_CSS_MODULES = "\0" + CSS_MODULES;

function getCssModulesCode(flightEnvironments: DevEnvironment[]) {
  const ids = new Set<string>();

  for (const environment of flightEnvironments) {
    const { idToModuleMap } = environment.moduleGraph;

    for (const [id, module] of idToModuleMap.entries()) {
      if (
        isCssModule(id) &&
        // On development, map may contain unused modules. Ignore them.
        module.importers.size !== 0
      )
        ids.add(id);
    }
  }

  return Array.from(ids)
    .map((id) => `import "${id}";`)
    .join("\n");
}
