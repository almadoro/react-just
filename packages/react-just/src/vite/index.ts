import react from "@vitejs/plugin-react";
import type { PluginOption } from "vite";
import build from "./build";
import dev from "./dev";
import { ReactJustOptions } from "./types";
import useClient from "./use-client";
import useServer from "./use-server";

const FLIGHT_MIME_TYPE = "text/x-component";

export default function reactJust(options: ReactJustOptions): PluginOption {
  return [
    build({ app: options.app, flightMimeType: FLIGHT_MIME_TYPE }),
    dev({ app: options.app, flightMimeType: FLIGHT_MIME_TYPE }),
    react(),
    useClient(),
    useServer(),
  ];
}
