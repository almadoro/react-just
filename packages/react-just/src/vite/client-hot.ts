import { DevEnvironment, Plugin } from "vite";
import { CSS_MODULES } from "./css";
import { ENVIRONMENTS, isFlightEnvironment } from "./environments";
import { CLIENT_MODULES } from "./use-client";

export default function clientHot(): Plugin {
  return {
    name: "react-just:client-hot",
    configureServer(server) {
      watchClientReactChange(server.environments[ENVIRONMENTS.CLIENT]);
    },
    hotUpdate(ctx) {
      if (isFlightEnvironment(this.environment.name))
        onFlightHotUpdate(ctx.server.environments[ENVIRONMENTS.CLIENT]);
    },
    resolveId(id) {
      switch (id) {
        case CLIENT_HOT_MODULES:
          return RESOLVED_CLIENT_HOT_MODULES;
        case BROWSER_HASH_CHECK:
          return RESOLVED_BROWSER_HASH_CHECK;
        case HMR_PREAMBLE:
          return RESOLVED_HMR_PREAMBLE;
        case SERVER_HMR:
          return RESOLVED_SERVER_HMR;
      }
    },
    load(id) {
      switch (id) {
        case RESOLVED_CLIENT_HOT_MODULES:
          return getClientHotModulesCode();
        case RESOLVED_BROWSER_HASH_CHECK:
          if (!(this.environment instanceof DevEnvironment)) return "";
          const browserHash =
            this.environment.depsOptimizer?.metadata.browserHash;
          if (!browserHash)
            throw new Error("Expected browser hash to be available");
          return getBrowserHashCheckCode(browserHash);
        case RESOLVED_HMR_PREAMBLE:
          return getHmrPreambleCode();
        case RESOLVED_SERVER_HMR:
          return getServerHmrCode();
      }
    },
  };
}

const BROWSER_HASH_CHECK_HOT_EVENT = "react-just:browser-hash-check";

function watchClientReactChange(clientEnv: DevEnvironment) {
  let currentBrowserHash = clientEnv.depsOptimizer?.metadata.browserHash;

  // When a reoptimization occurs, the React version of loaded modules on
  // client will not match future loaded ones. A full reload is required.
  setInterval(() => {
    let prevBrowserHash = currentBrowserHash;
    currentBrowserHash = clientEnv.depsOptimizer?.metadata.browserHash;

    const needsReload = prevBrowserHash !== currentBrowserHash;
    if (needsReload) clientEnv.hot.send({ type: "full-reload" });
  }, 0);

  // A reoptimization may occur before the Vite connection is established;
  // thus, any triggered full-reload events won't be received. A manual check
  // is required after the connection is established.
  clientEnv.hot.on("connection", () => {
    // The event is sent to every connected client. This doesn't matter.
    // This event should have no effects when the browser's modules are up to
    // date.
    clientEnv.hot.send({
      type: "custom",
      event: BROWSER_HASH_CHECK_HOT_EVENT,
      data: { browserHash: currentBrowserHash },
    });
  });
}

const TREE_RELOAD_HOT_EVENT = "react-just:tree-reload";

function onFlightHotUpdate(clientEnv: DevEnvironment) {
  // Send an event to trigger the client modules and react tree reload.
  clientEnv.hot.send({
    type: "custom",
    event: TREE_RELOAD_HOT_EVENT,
    data: { eventId: crypto.randomUUID() },
  });
}

export const CLIENT_HOT_MODULES = "/virtual:react-just/client-hot-modules";
const RESOLVED_CLIENT_HOT_MODULES = "\0" + CLIENT_HOT_MODULES;

function getClientHotModulesCode() {
  return (
    `import "${BROWSER_HASH_CHECK}";` +
    `import "${HMR_PREAMBLE}";` +
    `import "${SERVER_HMR}";`
  );
}

const BROWSER_HASH_CHECK = "/virtual:react-just/browser-hash-check";
const RESOLVED_BROWSER_HASH_CHECK = "\0" + BROWSER_HASH_CHECK;

function getBrowserHashCheckCode(currentBrowserHash: string) {
  return (
    `if (import.meta.hot) {` +
    `  import.meta.hot.on("${BROWSER_HASH_CHECK_HOT_EVENT}", ({ browserHash }) => {` +
    `    if (browserHash !== "${currentBrowserHash}")` +
    `      window.location.reload();` +
    `  });` +
    `}`
  );
}

const HMR_PREAMBLE = "/virtual:react-just/hmr-preamble";
const RESOLVED_HMR_PREAMBLE = "\0" + HMR_PREAMBLE;

function getHmrPreambleCode() {
  // Taken from: https://github.com/vitejs/vite-plugin-react/blob/main/packages/common/refresh-utils.ts
  return (
    `import { injectIntoGlobalHook } from "/@react-refresh";` +
    `if (import.meta.hot) {` +
    `  injectIntoGlobalHook(window);` +
    `  window.$RefreshReg$ = () => {};` +
    `  window.$RefreshSig$ = () => (type) => type;` +
    `}`
  );
}

const SERVER_HMR = "/virtual:react-just/server-hmr";
const RESOLVED_SERVER_HMR = "\0" + SERVER_HMR;

function getServerHmrCode() {
  return (
    `import { createFromRscFetch, render, RSC_MIME_TYPE } from "react-just/client";` +
    `if (import.meta.hot) {` +
    `  let lastEventId = null;` +
    `  import.meta.hot.on("${TREE_RELOAD_HOT_EVENT}", ({ eventId }) => {` +
    `    lastEventId = eventId;` +
    `    const headers = { accept: RSC_MIME_TYPE };` +
    // Don't use exactly the same URL as the one we're trying to load to avoid
    // sharing cache between the RSC and the HTML.
    `    const url = new URL(window.location.href);` +
    `    url.searchParams.set("__rsc__", "1");` +
    `    createFromRscFetch(fetch(url, { headers })).then(async (tree) => {` +
    // Add a timestamp to trigger modules reload on the browser.
    `      await import(/* @vite-ignore */ "${CLIENT_MODULES}?t=" + Date.now());` +
    `      await import(/* @vite-ignore */ "${CSS_MODULES}?t=" + Date.now());` +
    // Avoid race conditions between multiple reloads. Render only the latest one.
    `      if (lastEventId === eventId) render(tree);` +
    `    });` +
    `  });` +
    `}`
  );
}
