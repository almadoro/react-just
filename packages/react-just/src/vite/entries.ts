import { AppProps } from "@/types";
import { renderToPipeableStream as renderToPipeableHtmlStream } from "@/types/fizz.node";
import {
  decodePayloadIncomingMessage,
  renderToPipeableStream as renderToPipeableRscStream,
} from "@/types/flight.node";
import fs from "node:fs/promises";
import path from "node:path";
import type React from "react";
import { ComponentType } from "react";
import { Plugin } from "vite";
import { CLIENT_HOT_MODULES } from "./client-hot";
import { CSS_MODULES } from "./css";
import { ENVIRONMENTS } from "./environments";
import { CLIENT_MODULES, RESOLVED_CLIENT_MODULES } from "./use-client";
import { SERVER_FUNCTIONS_MODULES } from "./use-server";

type EntriesOptions = {
  app?: string;
};

export default function entries(options: EntriesOptions): Plugin {
  let appEntryId: string;

  return {
    name: "react-just:entries",
    async config(config) {
      const root = config.root ?? process.cwd();

      appEntryId = await resolveAppEntryId(root, options.app);

      return {
        root,
        environments: {
          [ENVIRONMENTS.CLIENT]: {
            optimizeDeps: {
              include: [
                "react",
                "react/jsx-runtime",
                "react/jsx-dev-runtime",
                "react-just/client",
              ],
            },
          },
          [ENVIRONMENTS.FIZZ_NODE]: {
            optimizeDeps: {
              include: [
                "react",
                "react/jsx-runtime",
                "react/jsx-dev-runtime",
                "react-just/fizz.node",
              ],
            },
          },
          [ENVIRONMENTS.FLIGHT_NODE]: {
            optimizeDeps: {
              entries: [appEntryId],
              include: [
                "react",
                "react/jsx-runtime",
                "react/jsx-dev-runtime",
                "react-just/flight.node",
              ],
            },
          },
        },
      };
    },
    resolveId(id) {
      switch (id) {
        case FLIGHT_ENTRY_NODE:
          return RESOLVED_FLIGHT_ENTRY_NODE;
        case FIZZ_ENTRY_NODE:
          return RESOLVED_FIZZ_ENTRY_NODE;
        case CLIENT_ENTRY:
          return RESOLVED_CLIENT_ENTRY;
        case SCAN_USE_CLIENT_ENTRY:
          return appEntryId;
        case SCAN_USE_SERVER_ENTRY:
          return RESOLVED_CLIENT_MODULES;
      }
    },
    async load(id) {
      switch (id) {
        case RESOLVED_FLIGHT_ENTRY_NODE:
          return getFlightEntry(appEntryId);
        case RESOLVED_FIZZ_ENTRY_NODE:
          return getFizzEntry();
        case RESOLVED_CLIENT_ENTRY:
          return getClientEntry();
      }
    },
  };
}

async function resolveAppEntryId(root: string, app?: string) {
  if (app) return toJsPath(path.resolve(root, app));

  for (const entryPath of APP_ENTRY_PATHS) {
    const entry = path.resolve(root, entryPath);
    try {
      await fs.access(entry, fs.constants.F_OK);
      return toJsPath(entry);
    } catch {}
  }

  throw new Error(
    `App entry not found. Specify the entry path with the "app" option or ` +
      `create a file named one of the following: ${APP_ENTRY_PATHS.join(", ")}`,
  );
}

function toJsPath(path: string) {
  return path.replace(/\\/g, "/");
}

const APP_ENTRY_PATHS = [
  "src/index.tsx",
  "src/index.jsx",
  "src/index.ts",
  "src/index.js",
];

export const FLIGHT_ENTRY_NODE = "/virtual:react-just/flight-entry.node";
const RESOLVED_FLIGHT_ENTRY_NODE = "\0" + FLIGHT_ENTRY_NODE;

function getFlightEntry(appEntryId: string) {
  return (
    `import React from "react";` +
    `import { decodePayloadIncomingMessage, renderToPipeableStream } from "react-just/flight.node";` +
    `import App from "${appEntryId}";` +
    `import "${SERVER_FUNCTIONS_MODULES}";` +
    `export { App, decodePayloadIncomingMessage, renderToPipeableStream, React };`
  );
}

export type FlightEntryNodeModule = {
  App: ComponentType<AppProps>;
  decodePayloadIncomingMessage: typeof decodePayloadIncomingMessage;
  renderToPipeableStream: typeof renderToPipeableRscStream;
  React: typeof React;
};

export const FIZZ_ENTRY_NODE = "/virtual:react-just/fizz-entry.node";
const RESOLVED_FIZZ_ENTRY_NODE = "\0" + FIZZ_ENTRY_NODE;

function getFizzEntry() {
  return (
    `import { renderToPipeableStream } from "react-just/fizz.node";` +
    `import "${CLIENT_MODULES}";` +
    `export { renderToPipeableStream };`
  );
}

export type FizzEntryNodeModule = {
  renderToPipeableStream: typeof renderToPipeableHtmlStream;
};

export const CLIENT_ENTRY = "/virtual:react-just/client-entry";
const RESOLVED_CLIENT_ENTRY = "\0" + CLIENT_ENTRY;

function getClientEntry() {
  return (
    `import "${CLIENT_HOT_MODULES}";` +
    `import "${CSS_MODULES}";` +
    `import "${CLIENT_MODULES}";` +
    `import { hydrateFromWindowStream } from "react-just/client";` +
    `hydrateFromWindowStream();`
  );
}

export const SCAN_USE_CLIENT_ENTRY =
  "/virtual:react-just/scan-use-client-entry";
export const SCAN_USE_SERVER_ENTRY =
  "/virtual:react-just/scan-use-server-entry";
