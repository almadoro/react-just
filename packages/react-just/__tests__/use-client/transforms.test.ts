import useClientTransform, {
  TransformOptions,
} from "@/vite/use-client/transform";
import { generate } from "astring";
import { builders } from "estree-toolkit";
import fs from "node:fs/promises";
import path from "node:path";
import { parseAst } from "vite";
import { describe, expect, test } from "vitest";

// client-like transform options
const noTreeshakeOpts: TransformOptions = {
  getRegisterArguments: ({ exportName, implementationIdentifier }) => [
    builders.identifier(implementationIdentifier),
    builders.literal(exportName),
  ],
  registerClientReferenceSource: "pkg",
  treeshakeImplementation: false,
};

// flight-like transform options
const treeshakeOpts: TransformOptions = {
  getRegisterArguments: ({ exportName }) => [builders.literal(exportName)],
  registerClientReferenceSource: "pkg",
  treeshakeImplementation: true,
};

describe("transforms with no treeshake", transformTests(noTreeshakeOpts));

describe("transforms with treeshake", transformTests(treeshakeOpts));

function transformTests(options: TransformOptions) {
  return () => {
    describe("export default declarations", () => {
      test(
        "class declarations",
        transformTest("export-default-declaration/class-declaration.js"),
      );

      test(
        "function declarations",
        transformTest("export-default-declaration/function-declaration.js"),
      );

      test(
        "identifier",
        transformTest("export-default-declaration/identifier.js"),
      );

      test(
        "value declarations",
        transformTest("export-default-declaration/value-declaration.js"),
      );
    });

    describe("export named declarations", () => {
      test(
        "array pattern",
        transformTest("export-named-declaration/array-pattern.js"),
      );

      test(
        "assignment pattern",
        transformTest("export-named-declaration/assignment-pattern.js"),
      );

      test("class", transformTest("export-named-declaration/class.js"));

      test(
        "identifier",
        transformTest("export-named-declaration/identifier.js"),
      );

      test("function", transformTest("export-named-declaration/function.js"));

      test(
        "object pattern",
        transformTest("export-named-declaration/object-pattern.js"),
      );

      test(
        "rest element",
        transformTest("export-named-declaration/rest-element.js"),
      );
    });

    test(
      "export named from source",
      transformTest("export-named-from-source.js"),
    );

    test(
      "export named specifiers",
      transformTest("export-named-specifiers.js"),
    );

    test("export all declarations not supported", () => {
      expect(() =>
        transform("'use client'; export * from 'pkg';"),
      ).toThrowError();
    });
  };

  function transformTest(filepath: string) {
    return async () => {
      const code = await fs.readFile(
        path.resolve(import.meta.dirname, "fixtures", filepath),
        "utf-8",
      );

      const output = transform(code);

      expect(output).toMatchSnapshot();
    };
  }

  function transform(code: string) {
    const ast = parseAst(code);

    useClientTransform(ast, options);

    return generate(ast);
  }
}
