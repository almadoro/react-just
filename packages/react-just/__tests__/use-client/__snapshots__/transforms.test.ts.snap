// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforms with no treeshake > export default declarations > class declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
class A {}
const $$Ref$$default = $$registerClientReference$$(A, "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with no treeshake > export default declarations > function declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
function $$default$$() {}
const $$Ref$$default = $$registerClientReference$$($$default$$, "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with no treeshake > export default declarations > identifier 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
let a;
const $$Ref$$default = $$registerClientReference$$(a, "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with no treeshake > export default declarations > value declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$default$$ = 1;
const $$Ref$$default = $$registerClientReference$$($$default$$, "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with no treeshake > export named declarations > array pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
let [a, {b}, [c], d = 1, , ...e] = [];
const $$Ref$$a = $$registerClientReference$$(a, "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$(b, "b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerClientReference$$(c, "c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerClientReference$$(d, "d");
export {$$Ref$$d as d};
const $$Ref$$e = $$registerClientReference$$(e, "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms with no treeshake > export named declarations > assignment pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
let [a = 1, {b} = c, [d] = e] = [];
const $$Ref$$a = $$registerClientReference$$(a, "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$(b, "b");
export {$$Ref$$b as b};
const $$Ref$$d = $$registerClientReference$$(d, "d");
export {$$Ref$$d as d};
"
`;

exports[`transforms with no treeshake > export named declarations > class 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
class A {}
const $$Ref$$A = $$registerClientReference$$(A, "A");
export {$$Ref$$A as A};
"
`;

exports[`transforms with no treeshake > export named declarations > function 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
function a() {}
const $$Ref$$a = $$registerClientReference$$(a, "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms with no treeshake > export named declarations > identifier 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
let a;
const $$Ref$$a = $$registerClientReference$$(a, "a");
export {$$Ref$$a as a};
"
`;

exports[`transforms with no treeshake > export named declarations > object pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
let {a, b: {c}, d: [e], f = 1, ...g} = {};
const $$Ref$$a = $$registerClientReference$$(a, "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$(c, "c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerClientReference$$(e, "e");
export {$$Ref$$e as e};
const $$Ref$$f = $$registerClientReference$$(f, "f");
export {$$Ref$$f as f};
const $$Ref$$g = $$registerClientReference$$(g, "g");
export {$$Ref$$g as g};
"
`;

exports[`transforms with no treeshake > export named declarations > rest element 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
let [...a] = [];
let [...[b, c]] = [];
let [...{d: e}] = [];
const $$Ref$$a = $$registerClientReference$$(a, "a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$(b, "b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerClientReference$$(c, "c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerClientReference$$(e, "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms with no treeshake > export named from source 1`] = `
"import {a as $$a$$, b as $$c$$, b as $$d$$, default as $$default$$, default as $$e$$} from "pkg";
import {registerClientReference as $$registerClientReference$$} from "pkg";
let c = 1;
const $$Ref$$a = $$registerClientReference$$($$a$$, "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$($$c$$, "c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerClientReference$$($$d$$, "d");
export {$$Ref$$d as d};
const $$Ref$$default = $$registerClientReference$$($$default$$, "default");
export {$$Ref$$default as default};
const $$Ref$$e = $$registerClientReference$$($$e$$, "e");
export {$$Ref$$e as e};
"
`;

exports[`transforms with no treeshake > export named specifiers 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
let a;
let b;
let d;
const $$Ref$$a = $$registerClientReference$$(a, "a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$(b, "c");
export {$$Ref$$c as c};
const $$Ref$$default = $$registerClientReference$$(d, "default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with treeshake > export default declarations > class declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$default = $$registerClientReference$$("default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with treeshake > export default declarations > function declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$default = $$registerClientReference$$("default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with treeshake > export default declarations > identifier 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$default = $$registerClientReference$$("default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with treeshake > export default declarations > value declarations 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$default = $$registerClientReference$$("default");
export {$$Ref$$default as default};
"
`;

exports[`transforms with treeshake > export named declarations > array pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$a = $$registerClientReference$$("a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$("b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerClientReference$$("c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerClientReference$$("d");
export {$$Ref$$d as d};
const $$Ref$$e = $$registerClientReference$$("e");
export {$$Ref$$e as e};
"
`;

exports[`transforms with treeshake > export named declarations > assignment pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$a = $$registerClientReference$$("a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$("b");
export {$$Ref$$b as b};
const $$Ref$$d = $$registerClientReference$$("d");
export {$$Ref$$d as d};
"
`;

exports[`transforms with treeshake > export named declarations > class 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$A = $$registerClientReference$$("A");
export {$$Ref$$A as A};
"
`;

exports[`transforms with treeshake > export named declarations > function 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$a = $$registerClientReference$$("a");
export {$$Ref$$a as a};
"
`;

exports[`transforms with treeshake > export named declarations > identifier 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$a = $$registerClientReference$$("a");
export {$$Ref$$a as a};
"
`;

exports[`transforms with treeshake > export named declarations > object pattern 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$a = $$registerClientReference$$("a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$("c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerClientReference$$("e");
export {$$Ref$$e as e};
const $$Ref$$f = $$registerClientReference$$("f");
export {$$Ref$$f as f};
const $$Ref$$g = $$registerClientReference$$("g");
export {$$Ref$$g as g};
"
`;

exports[`transforms with treeshake > export named declarations > rest element 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$a = $$registerClientReference$$("a");
export {$$Ref$$a as a};
const $$Ref$$b = $$registerClientReference$$("b");
export {$$Ref$$b as b};
const $$Ref$$c = $$registerClientReference$$("c");
export {$$Ref$$c as c};
const $$Ref$$e = $$registerClientReference$$("e");
export {$$Ref$$e as e};
"
`;

exports[`transforms with treeshake > export named from source 1`] = `
"import {a as $$a$$, b as $$c$$, b as $$d$$, default as $$default$$, default as $$e$$} from "pkg";
import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$a = $$registerClientReference$$("a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$("c");
export {$$Ref$$c as c};
const $$Ref$$d = $$registerClientReference$$("d");
export {$$Ref$$d as d};
const $$Ref$$default = $$registerClientReference$$("default");
export {$$Ref$$default as default};
const $$Ref$$e = $$registerClientReference$$("e");
export {$$Ref$$e as e};
"
`;

exports[`transforms with treeshake > export named specifiers 1`] = `
"import {registerClientReference as $$registerClientReference$$} from "pkg";
const $$Ref$$a = $$registerClientReference$$("a");
export {$$Ref$$a as a};
const $$Ref$$c = $$registerClientReference$$("c");
export {$$Ref$$c as c};
const $$Ref$$default = $$registerClientReference$$("default");
export {$$Ref$$default as default};
"
`;
